diff --git a/lib/sbi/sbi_ecall.c b/lib/sbi/sbi_ecall.c
index ce021eb..427ab51 100644
--- a/lib/sbi/sbi_ecall.c
+++ b/lib/sbi/sbi_ecall.c
@@ -94,7 +94,7 @@ void sbi_ecall_unregister_extension(struct sbi_ecall_extension *ext)
 
 int sbi_ecall_handler(struct sbi_trap_regs *regs)
 {
-	int ret = 0;
+	uint64_t ret = 0;
 	struct sbi_ecall_extension *ext;
 	unsigned long extension_id = regs->a7;
 	unsigned long func_id = regs->a6;
@@ -117,12 +117,14 @@ int sbi_ecall_handler(struct sbi_trap_regs *regs)
 		trap.epc = regs->mepc;
 		sbi_trap_redirect(regs, &trap);
 	} else {
+/*
 		if (ret < SBI_LAST_ERR) {
 			sbi_printf("%s: Invalid error %d for ext=0x%lx "
 				   "func=0x%lx\n", __func__, ret,
 				   extension_id, func_id);
 			ret = SBI_ERR_FAILED;
 		}
+*/
 
 		/*
 		 * This function should return non-zero value only in case of
